moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp

```diff
@@ -385,4 +385,101 @@ std::shared_ptr<BaseConstraint> createOMPLConstraint(const moveit::core::RobotMo
     return nullptr;
   }
 }
-}  // namespace ompl_interface
\ No newline at end of file
+
+/******************************************
+ * Toolpath Constraint
+ * ****************************************/
+ToolPathConstraint::ToolPathConstraint(const moveit::core::RobotModelConstPtr& robot_model, const std::string& group,
+                                       unsigned int num_dofs, unsigned int steps)
+  : BaseConstraint(robot_model, group, num_dofs), num_steps_(steps)
+{
+  NearestNeighbors<geometry_msgs::msg::Pose>::DistanceFunction func =
+      std::bind(&poseDistance, std::placeholders::_1, std::placeholders::_2);
+
+  pose_nn_.setDistanceFunction(func);
+}
+
+void ToolPathConstraint::setPath(const moveit_msgs::msg::CartesianTrajectory& msg)
+{
+  // TODO: Should this be done in the constructor? What happens when we want to accept another msg?
+
+  // Clear previous path
+  pose_nn_.clear();
+
+  // Interpolate along waypoints and insert into pose_nn_;
+  const auto dt = 1.0 / static_cast<double>(num_steps_);
+  for (unsigned int i = 0; i < msg.points.size() - 1; i++)
+  {
+    // Insert waypoints
+    pose_nn_.add(msg.points.at(i).point.pose);
+
+    // Insert interpolated poses
+    auto step = dt;
+    for (unsigned int j = 0; j < num_steps_ - 1; j++)
+    {
+      const auto pose = poseInterpolator(msg.points.at(i).point.pose, msg.points.at(i + 1).point.pose, step);
+      pose_nn_.add(pose);
+      step += dt;
+    }
+  }
+
+  // Insert last waypoint
+  pose_nn_.add(msg.points.at(msg.points.size() - 1).point.pose);
+}
+
+
+void ToolPathConstraint::function(const Eigen::Ref<const Eigen::VectorXd>& joint_values, Eigen::Ref<Eigen::VectorXd> out) const
+{
+  // TODO: Add path constraint function
+  // Ideal: F(q_actual) - F(q_neared_pose) = 0
+
```

Comment:
* Use forward kinematics on joint_values to compute EEF pose `pose1`
* Use `pose_nn_` to find pose `pose2` that is nearest to `pose1`
* Return `poseDistance(pose1, pose2)`